sled-agent/src/rack_setup/service.rs

```diff
@@ -1084,6 +1066,113 @@ impl ServiceInner {
     }
 }
 
+/// The service plan describes all the zones that we will eventually
+/// deploy on each sled.  But we cannot currently just deploy them all
+/// concurrently.  We'll do it in a few stages, each corresponding to a
+/// version of each sled's configuration.
+///
+/// - version 1: no services running
+///              (We don't have to do anything for this.  But we do
+///              reserve this version number for "no services running" so
+///              that sled agents can begin with an initial, valid
+///              OmicronZonesConfig before they've got anything running.)
+/// - version 2: internal DNS only
+/// - version 3: internal DNS + NTP servers
+/// - version 4: internal DNS + NTP servers + CockroachDB
+/// - version 5: everything
+///
+/// At each stage, we're specifying a complete configuration of what
+/// should be running on the sled -- including this version number.
+/// And Sled Agents will reject requests for versions older than the
+/// one they're currently running.  Thus, the version number is a piece
+/// of global, distributed state.
+///
+/// For now, we hardcode the requests we make to use specific version
+/// numbers.
+struct DeployStepVersion;
+
+impl DeployStepVersion {
+    const V1_NOTHING: Generation = OmicronZonesConfig::INITIAL_GENERATION;
+    const V2_DNS_ONLY: Generation = Self::V1_NOTHING.next();
+    const V3_DNS_AND_NTP: Generation = Self::V2_DNS_ONLY.next();
+    const V4_COCKROACHDB: Generation = Self::V3_DNS_AND_NTP.next();
+    const V5_EVERYTHING: Generation = Self::V4_COCKROACHDB.next();
+}
+
+fn build_initial_blueprint_from_plan(
+    sled_plan: &SledPlan,
+    service_plan: &ServicePlan,
+) -> anyhow::Result<Blueprint> {
+    let internal_dns_version =
+        Generation::try_from(service_plan.dns_config.generation)
+            .context("invalid internal dns version")?;
+
+    let mut sled_configs = BTreeMap::new();
+    for sled_request in sled_plan.sleds.values() {
+        let sled_addr = get_sled_address(sled_request.body.subnet);
+        let sled_id = sled_request.body.id;
+        let entry = match sled_configs.entry(sled_id) {
+            btree_map::Entry::Vacant(entry) => entry,
+            btree_map::Entry::Occupied(_) => {
+                bail!("duplicate sled address deriving blueprint: {sled_addr}");
+            }
+        };
+        let sled_config =
+            service_plan.services.get(&sled_addr).with_context(|| {
+                format!(
+                    "missing services in plan for sled {sled_id} ({sled_addr})"
+                )
+            })?;
+        entry.insert(sled_config.clone());
+    }
+
+    Ok(build_initial_blueprint_from_sled_configs(
+        sled_configs,
+        internal_dns_version,
+    ))
+}
+
+pub(crate) fn build_initial_blueprint_from_sled_configs(
```

Comment:
This ended up much simpler than I expected it would be. Good to know that our blueprints aren't *that* far from the RSS outputs.

Reply:
Same! I was pleasantly surprised.